examples/water_sampling_hrex.py

```diff
@@ -0,0 +1,225 @@
+# Water sampling script that uses HREX to generate fully-interacting end-state distributions
+
+# enable for 2x slow down
+# from jax import config
+# config.update("jax_enable_x64", True)
+
+
+import argparse
+import sys
+from dataclasses import replace
+from typing import Iterable, List, Optional, cast
+
+import matplotlib.pyplot as plt
+import numpy as np
+from rdkit import Chem
+from water_sampling_common import DEFAULT_BB_RADIUS, compute_occupancy, get_initial_state, setup_forcefield
+
+from timemachine.constants import DEFAULT_TEMP
+from timemachine.fe.free_energy import (
+    MDParams,
+    SimulationResult,
+    Trajectory,
+    make_pair_bar_plots,
+    run_sims_bisection,
+    run_sims_hrex,
+)
+from timemachine.fe.plots import (
+    plot_hrex_replica_state_distribution,
+    plot_hrex_replica_state_distribution_convergence,
+    plot_hrex_replica_state_distribution_heatmap,
+    plot_hrex_swap_acceptance_rates_convergence,
+    plot_hrex_transition_matrix,
+)
+from timemachine.ff import Forcefield
+
+
+def estimate_relative_free_energy_hrex(
+    mol: Chem.rdchem.Mol,
+    water_pdb,
+    seed,
+    lambda_min,
+    lambda_max,
+    n_windows,
+    ff: Forcefield,
+    nb_cutoff: float,
+    md_params: MDParams,
+    n_frames_bisection: int,
+    n_frames_per_iter: int,
+) -> SimulationResult:
+    temperature = DEFAULT_TEMP
+
+    combined_prefix = "hrex"
+
+    def make_optimized_initial_state(lamb):
+        state, _, _ = get_initial_state(water_pdb, mol, ff, seed, nb_cutoff, lamb)
+        return state
+
+    print("Starting bi-section")
+
+    md_params_bisection = replace(md_params, n_frames=n_frames_bisection)
+    # results_bisection, frames_by_state, boxes_by_state, final_velocities_by_state = run_sims_bisection(
+    results_bisection, trajectories_by_state = run_sims_bisection(
+        [lambda_min, lambda_max],
+        make_optimized_initial_state,
+        md_params_bisection,  # optimize?
+        n_bisections=n_windows - 2,
+        temperature=temperature,
+        min_overlap=0.666,
+    )
+
+    def get_mean_final_barostat_volume_scale_factor(trajectories_by_state: Iterable[Trajectory]) -> Optional[float]:
+        scale_factors = [traj.final_barostat_volume_scale_factor for traj in trajectories_by_state]
+        if any(x is not None for x in scale_factors):
+            assert all(x is not None for x in scale_factors)
+            sfs = cast(List[float], scale_factors)  # implied by assertion but required by mypy
+            return float(np.mean(sfs))
+        else:
+            return None
```

Comment:
nit: duplicated with rbfe.py https://github.com/proteneer/timemachine/blob/71319ef09595b43c2c9ce5565c871bbcaa199983/timemachine/fe/rbfe.py#L738-L745

Reply:
I think I've addressed this. I've refactored the original `estimate_relative_free_energy_bisection_hrex` into an `impl` method instead, so instead we just call the `impl` directly and now we D.R.Y.